kingfisher_scrapy/spiders/uganda_releases.py

```diff
@@ -1,62 +1,57 @@
+        '2019-2020'
+    Bulk download documentation
+        https://gpp.ppda.go.ug/public/open-data/ocds/ocds-datasets
     """
     name = 'uganda_releases'
-    download_delay = 30  # to avoid API 429 error "too many request"
     custom_settings = {
-        'CONCURRENT_REQUESTS': 1,
+        # We cannot get the list of all the files from https://gpp.ppda.go.ug/public/open-data/ocds/ocds-datasets
+        # because the list is generated in the browser.
+        # To get all the files, we follow the pattern download?fy={0}-{1}&code=1, iterating the 'code' value until it
+        # returns HTTP 500 error FileNotFoundException. Therefore, we retry all codes in RETRY_HTTP_CODES except 500.
+        'RETRY_HTTP_CODES': filter(lambda status: status != 500, RETRY_HTTP_CODES),
     }
-
     # BaseSpider
     date_format = 'year'
-    default_from_date = '2017'
+    default_from_date = '2019'
 
     # SimpleSpider
     data_type = 'release_package'
 
-    # IndexSpider
-    page_count_pointer = '/data/last_page'
-    parse_list_callback = 'parse_page'
-
-    def start_requests(self):
-        url = 'https://gpppapi.com/adminapi/public/api/pdes'
-        yield scrapy.Request(url, meta={'file_name': 'page-1.json'}, callback=self.parse_list)
-
-    @handle_http_error
-    def parse_page(self, response):
-        pattern = 'https://gpppapi.com/adminapi/public/api/open-data/v1/releases/{tag}?fy={fy}&pde={pde}'
-
-        for pdes in response.json()['data']['data']:
-            for plans in pdes['procurement_plans']:
-                for tag in ('planning', 'tender', 'award', 'contract'):
-                    if self.from_date and self.until_date:
-                        start_year = int(plans['financial_year'].split('-')[0])
-                        if not (self.from_date.year <= start_year <= self.until_date.year):
-                            continue
-                    yield self.build_request(
-                        pattern.format(tag=tag, fy=plans['financial_year'], pde=plans['pde_id']),
-                        formatter=join(components(-1), parameters('fy', 'pde'))
-                    )
+    # PeriodicSpider
+    formatter = staticmethod(parameters('fy', 'code'))
+    pattern = 'https://gpp.ppda.go.ug/adminapi/public/api/open-data/v2/ocds/download?fy={0}-{1}&code=1'
+    start_requests_callback = 'build_next'
+
+    def build_next(self, response):
+        if response.status == 500:
+            return
+        if not self.is_http_success(response):
+            yield self.build_file_error_from_response(response)
+        else:
+            yield from super().parse(response)
+            next_code = int(get_parameter_value(response.request.url, 'code')) + 1
+            yield self.build_request(replace_parameters(response.request.url, code=next_code),
+                                     formatter=self.formatter, callback=self.build_next)
```

Comment:
Just a little simpler (build_next -> parse, next_code -> code).
```suggestion

    def parse(self, response):
        if response.status == 500:
            return
        if not self.is_http_success(response):
            yield self.build_file_error_from_response(response)
        else:
            yield from super().parse(response)
            code = int(get_parameter_value(response.request.url, 'code')) + 1
            yield self.build_request(replace_parameters(response.request.url, code=code),
                                     formatter=self.formatter)
```